gh-119127: functools.partial placeholders

dg-pb opened the pull request:
As I already had implementation I though PR might be helpful for others to see and evaluate.

From all the different extensions of `functools.partial` I think this one is the best. It is relatively simple and exposes all missing functionality. Other `partial` extensions that I have seen lack functionality and would not provide complete argument ordering capabilities and/or are too complicated in relation to what they offer.

Implementation can be summarised as follows:
a) All trailing placeholders are removed. (Makes things simpler)
b) Throws exception if not all placeholders are filled on call
c) retains optimization benefits of application on other `partial` instances.

Performance penalty compared to current `functools.partial` is minimal for extension class. + 20-30 ns for initialisation and <4 ns when called with or without placeholders.

To put it simply, new functionality extends `functools.partial` so that it has flexibility of `lambda` / `def` approach (in terms of argument ordering), but call overhead is 2x smaller.

The way I see it is that this could only be justified if this extension provided completeness and no new functionality is going to be needed anywhere near in the future. I have thought about it and tried various alternatives and I think there is a good chance that this is the case. Personally, I don't think I would ever need anything more from `partial` class.

Current implementation functions reliably.

<!-- gh-issue-number: gh-119127 -->
* Issue: gh-119127
<!-- /gh-issue-number -->

dg-pb committed:
Initial Implementation

dg-pb committed:
Initial Implementation

dg-pb committed:
serialization fix

dg-pb committed:
serialization fix

dg-pb committed:
bug fix

dg-pb committed:
bug fix

nineteendo commented on Modules/_functoolsmodule.c:
```diff
@@ -41,13 +41,31 @@ get_functools_state(PyObject *module)
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_DISALLOW_INSTANTIATION,
+    .tp_new = PyType_GenericNew,
+};
```
https://github.com/python/cpython/actions/runs/9174174702/job/25224470051?pr=119303#step:16:69

> Non-constant global variables are generally not supported
in the CPython repo.

dg-pb replied:
I think it is because it didn't have `Py_TPFLAGS_IMMUTABLETYPE` flag, but can not check because the tool is not working for me localy.

nineteendo replied:
I think you can mark this as resolved.

dg-pb replied:
Finally! `c-analyzer` is not working for `clang` so could not run it locally.

dg-pb committed:
Bug 2 fix

dg-pb committed:
Bug 2 fix

dg-pb committed:
Py_TPFLAGS_IMMUTABLETYPE added

dg-pb committed:
Py_TPFLAGS_IMMUTABLETYPE added

dg-pb committed:
placeholder added to state as opposed to being used as global constant

dg-pb committed:
placeholder added to state as opposed to being used as global constant

dg-pb committed:
static removed

dg-pb committed:
static removed

dg-pb committed:
creating sentinel via PyType_Spec

dg-pb committed:
creating sentinel via PyType_Spec

dg-pb committed:
more accurate variable name

dg-pb committed:
more accurate variable name

dg-pb committed:
trailing trim bug and tests

dg-pb committed:
trailing trim bug and tests

dg-pb committed:
Updated docs

dg-pb committed:
Updated docs

dg-pb committed:
blurb

dg-pb committed:
blurb

Commit:
Merge branch 'main' into implement-119127

dg-pb committed:
minor edit

dg-pb committed:
minor edit

dg-pb committed:
doc fix

dg-pb committed:
doc fix

JelleZijlstra committed:
gh-119180: Update the magic number (#119397)

PR #119321 added a comment about the magic number bump
but did not actually apply the new magic number.

tim-one committed:
gh-119105: Differ.compare is too slow [for degenerate cases] (#119492)

``_fancy_replace()`` is no longer recursive. and a single call does a worst-case linear number of ratio() computations instead of quadratic. This renders toothless a universe of pathological cases. Some inputs may produce different output, but that's rare, and I didn't find a case where the final diff appeared to be of materially worse quality. To the contrary, by refusing to even consider synching on lines "far apart", there was more easy-to-digest locality in the output.

serhiy-storchaka committed:
gh-111999: Fix the signature of str.format_map() (#119540)

hugovk committed:
FAQ: Add reference to Python version numbering scheme (#119225)

pablogsal committed:
gh-119548: Add a 'clear' command to the REPL (#119549)

pablogsal committed:
Add codeowners for PYREPL (#119550)

AlexWaygood committed:
gh-99180: Make `StackSummary.should_show_carets` private (#119554)

ericsnowcurrently committed:
gh-119560: Drop an Invalid Assert in PyState_FindModule() (gh-119561)

The assertion was added in gh-118532 but was based on the invalid assumption that PyState_FindModule() would only be called with an already-initialized module def.  I've added a test to make sure we don't make that assumption again.

barneygale committed:
GH-82805: Fix handling of single-dot file extensions in pathlib (#118952)

pathlib now treats "`.`" as a valid file extension (suffix). This brings
it in line with `os.path.splitext()`.

In the (private) pathlib ABCs, we add a new `ParserBase.splitext()` method
that splits a path into a `(root, ext)` pair, like `os.path.splitext()`.
This method is called by `PurePathBase.stem`, `suffix`, etc. In a future
version of pathlib, we might make these base classes public, and so users
will be able to define their own `splitext()` method to control file
extension splitting.

In `pathlib.PurePath` we add optimised `stem`, `suffix` and `suffixes`
properties that don't use `splitext()`, which avoids computing the path
base name twice.

nedbat committed:
docs: fix a few typos identified by codespell (#119516)

Wulian233 committed:
Update README and layout.html from 3.13 to 3.14 (#119539)

Co-authored-by: Nice Zombies <nineteendo19d0@gmail.com>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

AlexWaygood committed:
gh-119562: Remove AST nodes deprecated since Python 3.8 (#119563)

scoder committed:
gh-111997: Fix argument count for LINE event and clarify type of argument counts. (#119179)

AlexWaygood committed:
gh-119562: Remove unused private string constants from `ast.py` (#119576)

gaogaotiantian committed:
gh-102864: Add switching frame test for pdb (#119564)

jaraco committed:
Re-order imports to align with zipp 3.18.2 (#119587)

xyb committed:
Fix typos in HISTORY documentation (#119453)

Eclips4 committed:
``Include/internal/pycore_import.h``: Fix typo (#119586)

Fix typo

adiaholic committed:
gh-119467: Fix Py_buffer.format type and correct documentation typo (#119475)

savannahostrowski committed:
GH-117283: Add doc warning for `PyTuple_SetItem` refcount > 1 (#117916)

rffontenelle committed:
Docs: Add class role for IPV{4,6}Address and fix a typo (#118059)

Add class role for IPV{4,6}Address and fix a typo

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

hugovk committed:
Docs: Only install sphinx-autobuild for `make htmllive` (#119607)

gaogaotiantian committed:
gh-119580: Improve version added section for convenience variable (#119583)

hugovk committed:
Docs: Move inline JavaScript to own file to reduce duplication (#119541)

gvanrossum committed:
Withdraw most of my ownership in favor of Mark (#119611)

rhettinger committed:
Misc cleanups and wording improvements for the itertools docs (gh-119626)

ericsnowcurrently committed:
gh-119584: Fix test_import Failed Assertion (gh-119623)

The fix in gh-119561 introduced an assertion that doesn't hold true if any of the three new test extension modules are loaded more than once.  This is fine normally but breaks if the new test_check_state_first() is run more than once, which happens for refleak checking and with the regrtest --forever flag.  We fix that here by clearing each of the three modules after loading them.  We also tweak a check in _modules_by_index_check().

erlend-aasland committed:
gh-117398: Add multiphase support to _datetime (gh-119373)

This is minimal support.  Subinterpreters are not supported yet.  That will be addressed in a later change.

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

JelleZijlstra committed:
gh-119311: Fix name mangling with PEP 695 generic classes (#119464)

Fixes #119311. Fixes #119395.

xyb committed:
Fix typos in comments (#119645)

serhiy-storchaka committed:
gh-117557: Improve error messages when a string, bytes or bytearray of length 1 are expected (GH-117631)

sobolevn committed:
gh-116860: Remove outdated `test_parserhack` from `test_future` (#116861)

jkunimune committed:
[doc] Clarify the nature of the root logger in the `logging` documentation (GH-119440)

Co-authored-by: Vinay Sajip <vinay_sajip@yahoo.co.uk>

sobolevn committed:
gh-118824: Remove deprecated `master_open` and `slave_open` from `pty` (#118826)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

sobolevn committed:
gh-119659: Move `@no_rerun` to `test.support` (#119660)

vstinner committed:
gh-119396: Optimize unicode_repr() (#119617)

Use stringlib to specialize unicode_repr() for each string kind
(UCS1, UCS2, UCS4).

Benchmark:

+-------------------------------------+---------+----------------------+
| Benchmark                           | ref     | change2              |
+=====================================+=========+======================+
| repr('abc')                         | 100 ns  | 103 ns: 1.02x slower |
+-------------------------------------+---------+----------------------+
| repr('a' * 100)                     | 369 ns  | 369 ns: 1.00x slower |
+-------------------------------------+---------+----------------------+
| repr(('a' + squote) * 100)          | 1.21 us | 946 ns: 1.27x faster |
+-------------------------------------+---------+----------------------+
| repr(('a' + nl) * 100)              | 1.23 us | 907 ns: 1.36x faster |
+-------------------------------------+---------+----------------------+
| repr(dquote + ('a' + squote) * 100) | 1.08 us | 858 ns: 1.25x faster |
+-------------------------------------+---------+----------------------+
| Geometric mean                      | (ref)   | 1.16x faster         |
+-------------------------------------+---------+----------------------+

stroxler committed:
gh-119581: Add a test of InitVar with name shadowing (#119582)

encukou committed:
gh-117398: gh-119655: datetime: Init static state once & don't free it (GH-119662)

- While datetime uses global state, only initialize it once.
- While `capi` is static, don't free it (thanks @neonene in https://github.com/python/cpython/pull/119641/files#r1616710048)

JelleZijlstra committed:
gh-117865: Defer import of re in ast (#119546)

This is used only by ast.get_source_segment(), so it seems sensible to avoid importing it.

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

sobolevn committed:
gh-119011: `type.__type_params__` now return an empty tuple (#119296)

iritkatriel committed:
gh-119676: remove several pseudo instructions which are use only in codegen (#119677)

lysnikolaou committed:
gh-119118: Fix performance regression in tokenize module (#119615)

* gh-119118: Fix performance regression in tokenize module

- Cache line object to avoid creating a Unicode object
  for all of the tokens in the same line.
- Speed up byte offset to column offset conversion by using the
  smallest buffer possible to measure the difference.

Co-authored-by: Pablo Galindo <pablogsal@gmail.com>

brandtbucher committed:
GH-119476: Split _CHECK_FUNCTION_VERSION out of _CHECK_FUNCTION_EXACT_ARGS (GH-119510)

brandtbucher committed:
GH-119258: Handle STORE_ATTR_WITH_HINT in tier two (GH-119481)

erlend-aasland committed:
gh-119538: Add missing expat build dependencies (#119647)

xmltok_impl.c and xmltok_ns.c are _included_ in xmltok.c by the C
pre-processor.

ericsnowcurrently committed:
gh-117398: Convert datetime.IsoCalendarDate To A Heap Type (gh-119637)

This is the only static type in the module that we will not keep static.

JelleZijlstra committed:
gh-119443: Turn off from __future__ import annotations in REPL (#119493)

Eclips4 committed:
gh-119704: Fix reference leak in the ``Python/Python-tokenize.c`` (#119705)

skirpichev committed:
gh-119555: catch SyntaxError from compile() in the InteractiveColoredConsole (#119557)

skirpichev committed:
gh-119613: Use C99+ functions instead of Py_IS_NAN/INFINITY/FINITE (#119619)

vstinner committed:
gh-119661: Add _Py_SINGLETON() include in Argumenet Clinic (#119712)

When the _Py_SINGLETON() is used, Argument Clinic now adds an
explicit "pycore_runtime.h" include to get the macro. Previously, the
macro may or may not be included indirectly by another include.

iritkatriel committed:
gh-119689: generate stack effect metadata for pseudo instructions (#119691)

matthiasgoergens committed:
gh-97588: Align ctypes struct layout to GCC/MSVC (GH-97702)

Structure layout, and especially bitfields, sometimes resulted in clearly
wrong behaviour like overlapping fields. This fixes

Co-authored-by: Gregory P. Smith <gps@python.org>
Co-authored-by: Petr Viktorin <encukou@gmail.com>

iritkatriel committed:
gh-93554: add test for quickening of code in loops ending with conditional statement (#119485)

skirpichev committed:
gh-119613: Soft deprecate Py_IS_NAN/INFINITY/FINITE (#119701)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

Zac-HD committed:
CI: set correct working directory for Hypothesis cache (GH-119345)

Set cwd for Hypothesis database

vstinner committed:
gh-119273: Don't run test_ioctl in a process group (#119275)

Python test runner no longer runs tests using TTY (ex: test_ioctl) in
a process group (using setsid()). Previously, tests using TTY were
skipped.

Eclips4 committed:
gh-119655: Fix reference leak in the ``_datetimemodule.c`` (gh-119713)

zooba committed:
gh-119690: Adds Unicode support for named pipes in _winapi (GH-119717)

filiplajszczak committed:
gh-119721: Integrate documentation fixes into heapq module docstring. (gh-119722)

jaraco committed:
gh-118673: Remove shebang and executable bits from stdlib modules. (#119658)

* gh-118673: Remove shebang and executable bits from stdlib modules.

* Removed shebangs and exe bits on turtledemo scripts.

The setting was inappropriate for '__main__' and inconsistent across the other modules. The scripts can still be executed directly by invoking with the desired interpreter.

hugovk committed:
gh-93963: Remove deprecated names from importlib.abc (#119720)

Co-authored-by: Jason R. Coombs <jaraco@jaraco.com>

adiaholic committed:
gh-119260: Clarify is_dataclass Behavior for Subclasses in Documentation and Tests (#119480)

Co-authored-by: Carl Meyer <carl@oddbird.net>

wimglenn committed:
gh-119594: Improve pow(fraction.Fraction(), b, modulo) error message (#119593)

If one calls pow(fractions.Fraction, x, module) with modulo not None, the error message now says that the types are incompatible rather than saying pow only takes 2 arguments.  Implemented by having fractions.Fraction __pow__ accept optional modulo argument and return NotImplemented if not None.  pow() then raises with appropriate message.
---------

Co-authored-by: Mark Dickinson <dickinsm@gmail.com>

zooba committed:
gh-119070: Fix py.exe handling of /usr/bin/env commands missing extension (GH-119426)

colesbury committed:
gh-119525: Fix deadlock with `_PyType_Lookup` and the GIL (#119527)

The deadlock only affected the free-threaded build and only occurred
when the GIL was enabled at runtime. The `Py_DECREF(old_name)` call
might temporarily release the GIL while holding the type seqlock.
Another thread may spin trying to acquire the seqlock while holding the
GIL.

The deadlock occurred roughly 1 in ~1,000 runs of `pool_in_threads.py`
from `test_multiprocessing_pool_circular_import`.

barneygale committed:
GH-89727: Partially fix `shutil.rmtree()` recursion error on deep trees (#119634)

Make `shutil._rmtree_unsafe()` call `os.walk()`, which is implemented
without recursion.

`shutil._rmtree_safe_fd()` is not affected and can still raise a recursion
error.

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

barneygale committed:
GH-119169: Implement `pathlib.Path.walk()` using `os.walk()` (#119573)

For silly reasons, pathlib's generic implementation of `walk()` currently
resides in `glob._Globber`. This commit moves it into
`pathlib._abc.PathBase.walk()` where it really belongs, and makes
`pathlib.Path.walk()` call `os.walk()`.

zware committed:
subprocess docs: Fix semantically important typo (GH-119752)

GH-25416 accidentally replaced a reference to the *stderr* argument of
`subprocess.run` with a reference to the *stdin* argument.  *stdin* is
not affected by the `check_output` option.

barneygale committed:
GH-89727: Fix `os.fwalk()` recursion error on deep trees (#119638)

Implement `os.fwalk()` using a list as a stack to avoid emitting recursion
errors on deeply nested trees.

barneygale committed:
GH-89727: Fix FD leak on `os.fwalk()` generator finalization. (#119766)

Follow-up to 3c890b50. Ensure we `os.close()` open file descriptors when
the `os.fwalk()` generator is finalized.

encukou committed:
gh-118235: Move RAISE_SYNTAX_ERROR actions to invalid rules and make sure they stay there (GH-119731)

The Full Grammar specification in the docs omits rule actions, so grammar rules that raise a syntax error looked like valid syntax.
This was solved in ef940de by hiding those rules in the custom syntax highlighter.

This moves all syntax-error alternatives to invalid rules, adds a validator that ensures that actions containing RAISE_SYNTAX_ERROR are in invalid rules, and reverts the syntax highlighter hack.

Lincoln-developer committed:
gh-118055: Update the finder glossary entry (GH-118278)

shenanigansd committed:
gh-110383: Remove references to removed popen[234] (GH-112783)


Signed-off-by: Bradley Reynolds <bradley.reynolds@darbia.dev>

emmatyping committed:
gh-119336: Restore removed _PyLong_NumBits() function (#119418)

It is used by the pywin32 project.

SweetyAngel committed:
gh-119779: Fix pyporting howto docs (#119785)

iritkatriel committed:
gh-119786: create folder in cpython repo for internals documentation (#119787)

serhiy-storchaka committed:
gh-119791: Fix new Tkinter tests for wantobjects=0 (GH-119792)

PhotoImage.get() retruns a string instead of a 3-tuple of integers
in this case.

serhiy-storchaka committed:
gh-109218: Refactor tests for the complex() constructor (GH-119635)

* Share common classes.
* Use exactly representable floats and exact tests.
* Check the sign of zero components.
* Remove duplicated tests (mostly left after merging int and long).
* Reorder tests in more consistent way.
* Test more error messages.
* Add tests for missed cases.

colesbury committed:
gh-119729: Use 't' in pkg-config file name for free-threaded build (#119738)

For example, the free-threaded build now generates
`lib/pkgconfig/python-3.13t.pc` and the debug build generates
`lib/pkgconfig/python-3.13d.pc`.

serhiy-storchaka committed:
gh-109218: Improve documentation for the complex() constructor (GH-119687)

* Remove the equivalence with real+imag*1j which can be incorrect in corner
  cases (non-finite numbers, the sign of zeroes).
* Separately document the three roles of the constructor: parsing a string,
  converting a number, and constructing a complex from components.
* Document positional-only parameters of complex(), float(), int() and bool()
  as positional-only.
* Add examples for complex() and int().
* Specify the grammar of the string for complex().
* Improve the grammar of the string for float().
* Describe more explicitly the behavior when real and/or imag arguments are
  complex numbers. (This will be deprecated in future.)

hugovk committed:
Docs: `shutil.rmtree`'s `onerror` has no pending removal version (#118947)

serhiy-storchaka committed:
gh-109218: Deprecate weird cases in the complex() constructor (GH-119620)

* Passing a string as the "real" keyword argument is now an error;
  it should only be passed as a single positional argument.
* Passing a complex number as the "real" or "imag" argument is now deprecated;
  it should only be passed as a single positional argument.

DBJim committed:
gh-107262: Update Tkinter tests for Tcl/Tk 8.6.14 (GH-119322)

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

iritkatriel committed:
gh-119744: move a few functions from compile.c to flowgraph.c (#119745)

dg-pb committed:
better variable names and mini corrections

dg-pb committed:
better variable names and mini corrections. Rebase

AlexWaygood commented on Lib/test/test_typing.py:
Please revert changes that are unrelated to this PR's purpose

nineteendo commented on Lib/_pyrepl/simple_interact.py:
```diff
@@ -97,8 +97,8 @@ def runsource(self, source, filename="<input>", symbol="single"):
-                    self.showsyntaxerror(filename)
-                    return False
+                self.showsyntaxerror(filename)
+                return False
```
```suggestion
                    self.showsyntaxerror(filename)
                    return False
```

nineteendo commented on Lib/test/test_typing.py:
```diff
@@ -7345,7 +7345,7 @@ def test_no_generator_instantiation(self):
     def test_async_generator(self):
         async def f():
-             yield 42
+            yield 42
```
```suggestion
             yield 42
```

nineteendo reviewed (commented):
Revert unrelated changes, as demanded by @AlexWaygood.

dg-pb committed:
merge to main

dg-pb committed:
Merge remote-tracking branch 'origin/implement-119127' into implement-119127

dg-pb committed:
..
